Fix user MD and tags getting deleted when restoring

KillianG opened the pull request:
Issue: CLDSRV-408

bert-e commented:
# Hello killiang,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-408 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.7.25`

* `8.8.0`


Please check the `Fix Version/s` of CLDSRV-408, or the target
branch of this pull request.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` into
`development/8.7`
* [w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring](https://github.com/scality/cloudserver/compare/development/8.8...w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring) will be merged into `development/8.8` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.6`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

francoisferrand commented on lib/api/apiUtils/object/versioning.js:
```diff
@@ -453,6 +453,13 @@ function overwritingVersioning(objMD, metadataStoreParams) {
 
+    // We need to keep user metadata and tags
+    for (const key of Object.keys(objMD)) {
+        if (key.startsWith('x-amz-meta-')) {
```
must not copy the "restore count" though ; and we should take the chance to check if there are other "scality" user metadata which should not be copied...

bert-e commented:
# History mismatch



Merge commit #7b6acd07bb86909ed890e675e492dcfcfe832cd5 on the integration branch
`w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` is merging a branch which is neither the current
branch `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` nor the development branch
`development/8.8`.

It is likely due to a rebase of the branch `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` into
`development/8.7`
* [w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring](https://github.com/scality/cloudserver/compare/development/8.8...w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring) will be merged into `development/8.8` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.6`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

KillianG commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` into
`development/8.7`
* [w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring](https://github.com/scality/cloudserver/compare/development/8.8...w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring) will be merged into `development/8.8` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.6`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

bert-e commented:
# History mismatch



Merge commit #5024cac0e4a3e5b2877981053c5fef376ff82b31 on the integration branch
`w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` is merging a branch which is neither the current
branch `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` nor the development branch
`development/8.8`.

It is likely due to a rebase of the branch `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

KillianG commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` into
`development/8.7`
* [w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring](https://github.com/scality/cloudserver/compare/development/8.8...w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring) will be merged into `development/8.8` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.6`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

francoisferrand commented on lib/api/apiUtils/object/versioning.js:
```diff
@@ -434,6 +434,43 @@ function preprocessingVersioningDelete(bucketName, bucketMD, objectMD, reqVersio
+    const userMDNotToKeep = ['x-amz-meta-scal-s3-restore-attempt'];
+    // We need to keep user metadata and tags
+    Object.keys(objMD).forEach(key => {
+        if (key.startsWith('x-amz-meta-')  && !userMDNotToKeep.includes(key)) {
```
```suggestion
        if (key.startsWith('x-amz-meta-') && !userMDNotToKeep.includes(key)) {
```

francoisferrand commented on lib/api/apiUtils/object/versioning.js:
```diff
@@ -434,6 +434,43 @@ function preprocessingVersioningDelete(bucketName, bucketMD, objectMD, reqVersio
+ */
+function restoreMetadatas(objMD, metadataStoreParams) {
+    /* eslint-disable no-param-reassign */
+    const userMDNotToKeep = ['x-amz-meta-scal-s3-restore-attempt'];
```
NotToKeep → Skip ?

francoisferrand commented on lib/api/apiUtils/object/versioning.js:
```diff
@@ -434,6 +434,43 @@ function preprocessingVersioningDelete(bucketName, bucketMD, objectMD, reqVersio
+        if (key.startsWith('x-amz-meta-')  && !userMDNotToKeep.includes(key)) {
+            metadataStoreParams.metaHeaders[key] = objMD[key];
+        }
+        if (key.startsWith('content-') && !userMDNotToKeep.includes(key)) {
```
should not check for "userMDNotToKeep", since this is not a userMetadata...

francoisferrand commented on lib/api/apiUtils/object/versioning.js:
```diff
@@ -434,6 +434,43 @@ function preprocessingVersioningDelete(bucketName, bucketMD, objectMD, reqVersio
+    // We need to keep user metadata and tags
+    Object.keys(objMD).forEach(key => {
+        if (key.startsWith('x-amz-meta-')  && !userMDNotToKeep.includes(key)) {
+            metadataStoreParams.metaHeaders[key] = objMD[key];
```
if we reach here, no point checking later (i.e. add `else` or `continue`)

francoisferrand commented on lib/api/apiUtils/object/versioning.js:
```diff
@@ -434,6 +434,43 @@ function preprocessingVersioningDelete(bucketName, bucketMD, objectMD, reqVersio
+        }
+    });
+
+    if (objMD.replicationInfo?.status === 'COMPLETED') {
```
shouldn't we copy the replication status even in the other cases?

* `PENDING` should probably not happen (as data would not have been available)... though it may be that we keep retrying or just failed but did not yet update the status → maybe we can print a warning, but we should not affect this workflow
* `FAILED` can definitely happen, and we should not make it look like anyhing changed...

francoisferrand commented on lib/api/apiUtils/object/versioning.js:
```diff
@@ -434,6 +434,43 @@ function preprocessingVersioningDelete(bucketName, bucketMD, objectMD, reqVersio
+    }
+
+    if (objMD.acl) {
+        metadataStoreParams.oldAcl = objMD.acl;
```
why `oldAcl`, and not just `acl` like the other?
(i don't see an already used `acl` field in here, but maybe i missed it)

KillianG replied:
acl is a global variable in service.js because we require it
`const acl = require('./metadata/acl');`

I wouldn't be able to create a variable like this if the field was named acl
```
const { objectKey, authInfo, size, contentMD5, metaHeaders,
            contentType, cacheControl, contentDisposition, contentEncoding,
            expires, multipart, headers, overrideMetadata, log,
            lastModifiedDate, versioning, versionId, uploadId,
            tagging, taggingCopy, replicationInfo, defaultRetention,
            dataStoreName, creationTime, retentionMode, retentionDate,
            legalHold, originOp, updateMicroVersionId, archive, oldReplayId,
            deleteNullKey, oldAcl } = params;
```

That said, I can still name it acl and use it as `params.acl`

What do you prefer?

francoisferrand replied:
personally, I would prefer to keep the name of the field clean (and possibly using another name when importing the package, to avoid problems)

not sure what others think though, if anyone knows "good practice" for this case, please share!

francoisferrand commented on tests/unit/api/apiUtils/versioning.js:
```diff
@@ -616,4 +619,345 @@ describe('versioning helpers', () => {
+            {
+                description: 'Should update archive with restore infos',
+                objMD: {
+                'versionId': '2345678',
+                'creation-time': now,
+                'last-modified': now,
+                'originOp': 's3:PutObject',
+                archive: {
+                    'restoreRequestedDays': days,
+                    'restoreRequestedAt': now,
+                    archiveInfo
+                } },
+                expectedRes: {
```
missing indent:

```suggestion
                    'versionId': '2345678',
                    'creation-time': now,
                    'last-modified': now,
                    'originOp': 's3:PutObject',
                    archive: {
                        'restoreRequestedDays': days,
                        'restoreRequestedAt': now,
                        archiveInfo
                    },
                },
                expectedRes: {
```

francoisferrand reviewed (commented):
Due to node.js "untyped" quality, the unit test do not ensure the code works : need to add extra "functional" tests in `scality/cloudserver/tests/functional/aws-node-sdk/test/object/putVersion.js`, which will ensure that:
* update versioning is indeed called
* the objMD it gets is indeed correctly filled
* the resulting "metadataObject" has the field which match what is required by `metadataStoreObject`

bert-e commented:
# History mismatch



Merge commit #319a4428c1233805362eb9d7cc9da5a9392a8fd7 on the integration branch
`w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` is merging a branch which is neither the current
branch `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` nor the development branch
`development/8.8`.

It is likely due to a rebase of the branch `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

KillianG commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` into
`development/8.7`
* [w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring](https://github.com/scality/cloudserver/compare/development/8.8...w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring) will be merged into `development/8.8` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.6`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

francoisferrand commented on lib/services.js:
```diff
@@ -128,12 +128,11 @@ const services = {
+        if (cipherBundle || params.cipherBundle) {
+            const bundle = cipherBundle || params.cipherBundle;
+            md.setAmzServerSideEncryption(bundle.algorithm);
+            md.setAmzEncryptionKeyId(bundle.configuredMasterKeyId || bundle.masterKeyId);
```
should we not keep the test which was present before?
e.g. do not set `AmzEncryptionKeyId` to null-ish value?

KazToozs commented on tests/unit/api/apiUtils/versioning.js:
```diff
@@ -616,4 +619,346 @@ describe('versioning helpers', () => {
+                    taggingCopy: { 'testtag': 'testtag', 'testtag2': 'testtag2' },
+                    archive: {
+                        archiveInfo,
+                        'restoreRequestedDays': 3,
```
could we not use the `days` variable here?

KazToozs commented on tests/unit/api/apiUtils/versioning.js:
```diff
@@ -616,4 +619,346 @@ describe('versioning helpers', () => {
+                },
+            },
+            {
+                description: 'Should return nullVersionId',
```
I'm not sure this test title is explicit enough

francoisferrand reviewed (changes requested):
Code change looks good (just a few details, c.f. comments), but missing functional tests

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers






The following reviewers are expecting changes from the author, or must review again:

* @francoisferrand

nicolas2bert commented on lib/services.js:
```diff
@@ -128,12 +128,11 @@ const services = {
-                md.setAmzEncryptionKeyId(cipherBundle.configuredMasterKeyId || cipherBundle.masterKeyId);
-            }
+
+        if (cipherBundle || params.cipherBundle) {
```
Why don't we always use the `cipherBundle` generated from the request header?
Using the "copied" `params.cipherBundle` creates data discrepancy because the data is being encrypted with the cipher bundle generated from the request header.

nicolas2bert commented on lib/api/apiUtils/object/versioning.js:
```diff
@@ -434,6 +434,57 @@ function preprocessingVersioningDelete(bucketName, bucketMD, objectMD, reqVersio
+        metadataStoreParams.replicationInfo = objMD.replicationInfo;
+    }
+
+    if (objMD.legalHold) {
```
What about the retention date and mode?

KillianG replied:
Retention date and mode are bucket level so we don't need to manage them here

nicolas2bert commented on tests/functional/aws-node-sdk/test/utils/init.js:
```diff
@@ -79,6 +79,31 @@ function fakeMetadataArchive(bucketName, objectName, versionId, archive, cb) {
         /* eslint-disable no-param-reassign */
         objMD.dataStoreName = 'location-dmf-v1';
         objMD.archive = archive;
+        objMD['x-amz-meta-custom-user-md'] = 'custom-md';
```
acl, legalHold (and potentially retention date and mode) not tested?

nicolas2bert commented on lib/api/apiUtils/object/versioning.js:
```diff
@@ -434,6 +434,57 @@ function preprocessingVersioningDelete(bucketName, bucketMD, objectMD, reqVersio
+        if (key.startsWith('x-amz-meta-') && !userMDToSkip.includes(key)) {
+            metadataStoreParams.metaHeaders[key] = objMD[key];
+        } else if (key.startsWith('content-')) {
+            metadataStoreParams[key] = objMD[key];
```
when is `metadataStoreParams['content-xxx']` used?

nicolas2bert commented on lib/api/apiUtils/object/versioning.js:
```diff
@@ -434,6 +434,57 @@ function preprocessingVersioningDelete(bucketName, bucketMD, objectMD, reqVersio
+        }
+    });
+
+    if (objMD['x-amz-server-side-encryption']) {
```
We should not "copy" any metadata that is data-related (data encryption, data size...).

bert-e commented:
# History mismatch



Merge commit #4f4762f902f7ed9fa026375dac5f147f8ed0f0c5 on the integration branch
`w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` is merging a branch which is neither the current
branch `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` nor the development branch
`development/8.8`.

It is likely due to a rebase of the branch `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

KillianG commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` into
`development/8.7`
* [w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring](https://github.com/scality/cloudserver/compare/development/8.8...w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring) will be merged into `development/8.8` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.6`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers






The following reviewers are expecting changes from the author, or must review again:

* @francoisferrand

bert-e commented:
# History mismatch



Merge commit #e7730a41ea96bd61da07c393c1d42d271d2df360 on the integration branch
`w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` is merging a branch which is neither the current
branch `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` nor the development branch
`development/8.8`.

It is likely due to a rebase of the branch `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

KillianG commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` into
`development/8.7`
* [w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring](https://github.com/scality/cloudserver/compare/development/8.8...w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring) will be merged into `development/8.8` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.6`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers






The following reviewers are expecting changes from the author, or must review again:

* @francoisferrand

bert-e commented:
# History mismatch



Merge commit #d4a2e5483def5fd4edff15271e2b2e2d3dc6d2cd on the integration branch
`w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` is merging a branch which is neither the current
branch `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` nor the development branch
`development/8.8`.

It is likely due to a rebase of the branch `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

KillianG commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` into
`development/8.7`
* [w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring](https://github.com/scality/cloudserver/compare/development/8.8...w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring) will be merged into `development/8.8` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.6`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers






The following reviewers are expecting changes from the author, or must review again:

* @francoisferrand

nicolas2bert commented:
Much cleaner. 
Thanks. 
Last request for me, can we add tests for `should not "copy" any metadata that is data-related (data encryption, data size...)`?

francoisferrand commented on tests/functional/aws-node-sdk/test/object/putVersion.js:
```diff
@@ -89,8 +89,13 @@ describe('PUT object with x-scal-s3-version-id header', () => {
+                        return next(err);
+                    }
+                    // eslint-disable-next-line no-param-reassign
+                    objMD.legalHold = true;
```
why add 'legalHold' to each object?
* we probably want to the test both cases (overwrite with and without legal hold)
* setting legal hold should be done via regular s3 api, not by tweaking the API

francoisferrand commented on tests/functional/aws-node-sdk/test/object/putVersion.js:
```diff
@@ -89,8 +89,13 @@ describe('PUT object with x-scal-s3-version-id header', () => {
+                    objMD.legalHold = true;
                     objMDBefore = objMD;
-                    return next(err);
+                    return metadata.putObjectMD(bucketName, objectName, objMD, params, log, err => next(err));
```
should not overwrite metadata manually, this breaks the semantics of the scenario, no?

bert-e commented:
# History mismatch



Merge commit #2a17f0d4b39121578e9292eac31d45a906cd6237 on the integration branch
`w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` is merging a branch which is neither the current
branch `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` nor the development branch
`development/8.8`.

It is likely due to a rebase of the branch `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Kerkesni commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` into
`development/8.7`
* [w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring](https://github.com/scality/cloudserver/compare/development/8.8...w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring) will be merged into `development/8.8` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.6`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers






The following reviewers are expecting changes from the author, or must review again:

* @francoisferrand

francoisferrand commented on tests/functional/aws-node-sdk/test/object/mpuVersion.js:
```diff
@@ -847,5 +847,87 @@ describe('MPU with x-scal-s3-version-id header', () => {
+                    objMD['content-encoding'] = 'testencoding';
+                    objMD['x-amz-server-side-encryption'] = 'aws:kms';
+                    /* eslint-enable no-param-reassign */
+                    return metadata.putObjectMD(bucketName, objectName, objMD, undefined, log, next);
```
can we use the proper AWS api to set the metadata fields?

this is _functional_ test (not unit test), so it would be better to reproduce the 'real' use case... (though we may indeed putObjectMD to fake the 'archival' of the object)

francoisferrand commented on tests/functional/aws-node-sdk/test/object/objectCopy.js:
```diff
@@ -1271,25 +1279,36 @@ describe('Object Copy', () => {
+                    /* eslint-disable no-param-reassign */
+                    objMD.dataStoreName = 'location-dmf-v1';
+                    objMD.archive = archiveCompleted;
+                    objMD['x-amz-meta-custom-user-md'] = 'custom-md';
```
same here

francoisferrand commented on tests/functional/aws-node-sdk/test/object/putVersion.js:
```diff
@@ -753,5 +753,88 @@ describe('PUT object with x-scal-s3-version-id header', () => {
+                    objMD['content-md5'] = 'testmd 5';
+                    objMD['content-encoding'] = 'testencoding';
+                    objMD['x-amz-server-side-encryption'] = 'aws:kms';
+                    /* eslint-enable no-param-reassign */
```
same

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

Commit:
keep same object properties after a restore of a cold object

Object properties such as ACLs and custom user metadata should
not be removed after the restore of a cold object.

Issue: CLDSRV-408

bert-e commented:
# History mismatch



Merge commit #f8a54b12dc1a5077af8f33853da1e6588b400d24 on the integration branch
`w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` is merging a branch which is neither the current
branch `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` nor the development branch
`development/8.8`.

It is likely due to a rebase of the branch `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Kerkesni commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` with contents from `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring`
and `development/8.8`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring origin/development/8.8
 $ git merge origin/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

bert-e commented:
# History mismatch



Merge commit #77164631df4395f2cb5f16ba15c50a7751763738 on the integration branch
`w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` is merging a branch which is neither the current
branch `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` nor the development branch
`development/8.8`.

It is likely due to a rebase of the branch `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Commit:
add tests for keeping object properties after restore

Issue: CLDSRV-408

Commit:
bump version to 8.7.26

Kerkesni commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring` with contents from `improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring`
and `development/8.8`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring origin/development/8.8
 $ git merge origin/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.8/improvement/CLDSRV-408-Fix-metadata-getting-deleted-when-restoring
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

francoisferrand commented:
/bypass_author_approval

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.7`

* :heavy_check_mark: `development/8.8`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.6`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **bypass_author_approval**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.7`

* :heavy_check_mark: `development/8.8`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.6`



Please check the status of the associated issue CLDSRV-408.

Goodbye killiang.
